prusti-specs/src/rewriter.rs

```diff
@@ -116,70 +96,105 @@ impl AstRewriter {
             let fn_arg = self.generate_result_arg(item);
             spec_item.sig.inputs.push(fn_arg);
         }
         Ok(syn::Item::Fn(spec_item))
     }
 
-    /// Generate statements for checking the given loop invariant.
-    pub fn generate_spec_loop(
+    /// Parse an assertion into a Rust expression
+    pub fn process_assertion(
+        &mut self,
+        spec_type: SpecItemType,
+        spec_id: untyped::SpecificationId,
+        tokens: TokenStream,
+        item: &untyped::AnyFnItem,
+    ) -> syn::Result<syn::Item> {
+        self.generate_spec_item_fn(
+            spec_type,
+            spec_id,
+            Parser::from_token_stream(tokens).extract_assertion_expr()?,
+            item
+        )
+    }
+
+    /// Parse a pledge without lhs into a Rust expression
+    pub fn process_pledge_rhs_only(
+        &mut self,
+        spec_id: untyped::SpecificationId,
+        tokens: TokenStream,
+        item: &untyped::AnyFnItem,
+    ) -> syn::Result<syn::Item> {
+        self.generate_spec_item_fn(
+            SpecItemType::Pledge,
+            spec_id,
+            Parser::from_token_stream(tokens).extract_pledge_rhs_only_expr()?,
+            &item
+        )
+    }
+
+    /// Parse a pledge with lhs into a Rust expression
+    pub fn process_pledge(
+        &mut self,
+        spec_id: untyped::SpecificationId,
+        tokens: TokenStream,
+        item: &untyped::AnyFnItem,
+    ) -> syn::Result<syn::Item> {
+        self.generate_spec_item_fn(
+            SpecItemType::Pledge,
+            spec_id,
+            Parser::from_token_stream(tokens).extract_pledge_expr()?,
+            &item
+        )
+    }
+
+    /// Parse a loop invariant into a Rust expression
+    pub fn process_loop(
         &mut self,
         spec_id: untyped::SpecificationId,
-        assertion: untyped::Assertion,
-    ) -> TokenStream {
-        let mut statements = TokenStream::new();
-        assertion.encode_type_check(&mut statements);
+        tokens: TokenStream,
+    ) -> syn::Result<TokenStream> {
+        let expr = Parser::from_token_stream(tokens).extract_assertion_expr()?;
         let spec_id_str = spec_id.to_string();
-        let assertion_json = crate::specifications::json::to_json_string(&assertion);
         let callsite_span = Span::call_site();
-        quote_spanned! {callsite_span=>
+        Ok(quote_spanned! {callsite_span=>
             #[allow(unused_must_use, unused_variables)]
             {
                 #[prusti::spec_only]
                 #[prusti::loop_body_invariant_spec]
                 #[prusti::spec_id = #spec_id_str]
-                #[prusti::assertion = #assertion_json]
-                || {
-                    #statements
+                || -> bool {
+                    #expr
                 };
             }
-        }
+        })
     }
 
-    /// Generate statements for checking a closure specification.
-    /// TODO: arguments, result (types are typically not known yet after parsing...)
```

Comment:
Please keep the TODO.

Reply:
Fixed in 681ccfe0b2aa35aa9ed35a05b55490ba69ad34b9